File: ai/schemas/persona.py
"""Pydantic mirror of `shared/schemas/persona.ts`."""

from __future__ import annotations

from pydantic import Field

from .audience import AudienceSegment
from .base import ReelLabModel


class AttentionProfile(ReelLabModel):
    average_attention_seconds: float = Field(ge=0.0)
    swipe_tendency: float = Field(ge=0.0, le=1.0)
    drop_off_triggers: list[str] = Field(default_factory=list)


class EngagementProfile(ReelLabModel):
    like_tendency: float = Field(ge=0.0, le=1.0)
    save_tendency: float = Field(ge=0.0, le=1.0)
    share_tendency: float = Field(ge=0.0, le=1.0)
    comment_tendency: float = Field(ge=0.0, le=1.0)
    follow_tendency: float | None = Field(default=None, ge=0.0, le=1.0)


class DurationRange(ReelLabModel):
    min: float
    max: float


class ContentPreferences(ReelLabModel):
    preferred_formats: list[str] = Field(default_factory=list)
    preferred_tones: list[str] = Field(default_factory=list)
    preferred_duration_seconds: DurationRange
    turn_offs: list[str] = Field(default_factory=list)


class Persona(ReelLabModel):
    id: str
    segment_id: str
    name: str
    demographic_summary: str
    interests: list[str] = Field(default_factory=list)
    behavioral_traits: list[str] = Field(default_factory=list)
    attention_profile: AttentionProfile
    engagement_profile: EngagementProfile
    content_preferences: ContentPreferences
    system_brief: str | None = None


class PersonaGenerationRequest(ReelLabModel):
    segment: AudienceSegment
    count: int = Field(default=3, ge=1, le=25)
    creator_goal: str | None = None

File: shared/schemas/persona.ts
/**
 * ReelLab shared contract — Persona.
 *
 * OWNER OF CHANGES: Developer 1 (AI / Simulation Lead).
 * Types only — no runtime code. See shared/README.md for the contract rules.
 */

import type { AudienceSegment, UnitScore } from './audience';

/** How long a persona is willing to give a video before deciding to swipe. */
export interface AttentionProfile {
  /** Typical seconds before the persona decides to keep watching or swipe. */
  averageAttentionSeconds: number;
  /** How aggressively this persona swipes away, 0 (patient) .. 1 (ruthless). */
  swipeTendency: UnitScore;
  /** Things that reliably lose this persona, e.g. ["slow intro", "no captions"]. */
  dropOffTriggers: string[];
}

/** Baseline engagement rates, before any specific content is considered. */
export interface EngagementProfile {
  likeTendency: UnitScore;
  saveTendency: UnitScore;
  shareTendency: UnitScore;
  commentTendency: UnitScore;
  /** Chance this persona follows after a single strong reel, 0..1. */
  followTendency?: UnitScore;
}

/** Content this persona gravitates toward or bounces off. */
export interface ContentPreferences {
  preferredFormats: string[];
  preferredTones: string[];
  preferredDurationSeconds: { min: number; max: number };
  turnOffs: string[];
}

/**
 * A synthetic viewer. Personas are generated per `AudienceSegment` and are the
 * unit of simulation — one persona produces one `PersonaSimulationResult`.
 */
export interface Persona {
  id: string;
  /** The `AudienceSegment.id` this persona was generated for. */
  segmentId: string;
  name: string;
  /** One-line demographic sketch, e.g. "21, male, Pune, engineering student". */
  demographicSummary: string;
  interests: string[];
  behavioralTraits: string[];
  attentionProfile: AttentionProfile;
  engagementProfile: EngagementProfile;
  contentPreferences: ContentPreferences;
  /**
   * Compact natural-language brief handed to the reasoning model when this
   * persona watches a reel. Generated once, then cached — see the persona
   * caching note in docs/architecture.md.
   */
  systemBrief?: string;
}

/** Request body for persona generation. */
export interface PersonaGenerationRequest {
  segment: AudienceSegment;
  /** How many personas to generate for this segment. */
  count: number;
  /** Optional creator goal, so personas are generated with intent in mind. */
  creatorGoal?: string;
}
